File: mysky-api/src/main/java/com/mysky/api/entity/Post.java
package com.mysky.api.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Entity
@Table(name = "posts")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Post {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(length = 5000, nullable = false)
    private String content;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "author_id", nullable = false)
    private User author;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "circle_id")
    private Circle circle;

    @Enumerated(EnumType.STRING)
    @Builder.Default
    private Visibility visibility = Visibility.PUBLIC;

    @ElementCollection
    @CollectionTable(name = "post_tags", joinColumns = @JoinColumn(name = "post_id"))
    @Column(name = "tag")
    @Builder.Default
    private List<String> tags = new ArrayList<>();

    @ElementCollection
    @CollectionTable(name = "post_mentions", joinColumns = @JoinColumn(name = "post_id"))
    @Column(name = "user_id")
    @Builder.Default
    private Set<Long> mentions = new HashSet<>();

    @ElementCollection
    @CollectionTable(name = "post_media", joinColumns = @JoinColumn(name = "post_id"))
    @Column(name = "media_url")
    @Builder.Default
    private List<String> mediaUrls = new ArrayList<>();

    private Double latitude;
    private Double longitude;
    private String location;

    @ManyToMany
    @JoinTable(
        name = "post_likes",
        joinColumns = @JoinColumn(name = "post_id"),
        inverseJoinColumns = @JoinColumn(name = "user_id")
    )
    @Builder.Default
    private Set<User> likedBy = new HashSet<>();

    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<Comment> comments = new ArrayList<>();

    @CreationTimestamp
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;

    public enum Visibility {
        PUBLIC, FOLLOWERS, CIRCLE, PRIVATE
    }
}

File: My-sky-main/src/models/Post.ts
import { Timestamp } from '@react-native-firebase/firestore';

export interface Post {
  id: string;
  userId: string;
  text: string;
  image?: string;
  createdAt: Timestamp;
  updatedAt?: Timestamp;
  likes: string[]; // ユーザーIDの配列
  comments: number; // コメント数
  quotedPostId?: string; // 引用投稿のID
  replyToPostId?: string; // 返信先の投稿ID
  mentions?: string[]; // 言及されたユーザーIDの配列
  tags?: string[]; // ハッシュタグの配列
  location?: {
    latitude: number;
    longitude: number;
    name?: string;
  };
}

export interface PostCreation {
  text: string;
  image?: string;
  quotedPostId?: string;
  replyToPostId?: string;
  mentions?: string[];
  tags?: string[];
  location?: {
    latitude: number;
    longitude: number;
    name?: string;
  };
}

export interface Comment {
  id: string;
  postId: string;
  userId: string;
  text: string;
  createdAt: Timestamp;
  likes: string[]; // ユーザーIDの配列
}

export interface CommentCreation {
  postId: string;
  text: string;
}

export interface PostWithUser extends Post {
  user: {
    id: string;
    nickname: string;
    profilePhoto: string;
  };
  isLiked: boolean;
  blockedUsers?: string[]; // ユーザーがブロックしているユーザーIDのリスト
} 